### PR TITLE
Add the changes to handle upsert to remove document and context manag…

### PR DESCRIPTION
**Pull request by sunank200**
- Use context manager for batch ingestion in [weaviate](https://weaviate.io/developers/weaviate/manage-data/import).
- Fix the logic to remove successfully upserted documents.

closes: [#174](https://github.com/astronomer/ask-astro/issues/174)
closes: [#173](https://github.com/astronomer/ask-astro/issues/173)

### CODE REVIEW

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py [hunk 1]
```diff
@@ -220,84 +222,91 @@ def batch_ingest(
         :param uuid_column: Name of the column containing the UUID.
         :param vector_column: Name of the column containing the vector data.
         :param batch_params: Parameters for batch configuration.
-        :param existing: Strategy to handle existing data ('skip', 'replace', 'upsert').
-        :param verbose: Whether to print verbose output.
+        :param existing: Strategy to handle existing data ('skip', 'replace', 'upsert' or 'error').
+        :param verbose: Whether to log verbose output.
+        :param tenant: The tenant to which the object will be added.
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:227]: I would suggest we keep the docstring order consistent with the parameters

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py [hunk 2]
```diff
@@ -220,84 +222,91 @@ def batch_ingest(
...
+        with self.client.batch as batch:
+            for row_id, row in df.iterrows():
+                data_object = row.to_dict()
+                uuid = data_object.pop(uuid_column)
+                vector = data_object.pop(vector_column, None)
+
+                try:
+                    if self.client.data_object.exists(uuid=uuid, class_name=class_name):
+                        if existing == "error":
+                            raise AirflowException(f"Ingest of UUID {uuid} failed.  Object exists.")
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:245]:
```suggestion
                            raise AirflowException(f"Ingest of UUID {uuid} failed. Object exists.")

```

Comment (Lee-W) [reply airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:245]: add the additional line for separate the 2 if-else logic

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py [hunk 3]
```diff
@@ -220,84 +222,91 @@ def batch_ingest(
...
+
+                except Exception as e:
+                    if isinstance(e, AirflowException):
+                        self.logger.error(f"Failed to add row {row_id} with UUID {uuid}. Error: {e}")
+                        self.batch_errors.append({"uuid": uuid, "result": {"errors": str(e)}})
+                        break
+                    else:
+                        if verbose:
+                            self.logger.error(f"Failed to add row {row_id} with UUID {uuid}. Error: {e}")
+                        self.batch_errors.append({"uuid": uuid, "result": {"errors": str(e)}})
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:263]:
```suggestion
                except AirflowException as e:
                    self.logger.error(f"Failed to add row {row_id} with UUID {uuid}. Error: {e}")
                    self.batch_errors.append({"uuid": uuid, "result": {"errors": str(e)}})
                    break
                except Exception as e:
                    if verbose:
                        self.logger.error(f"Failed to add row {row_id} with UUID {uuid}. Error: {e}")
                    self.batch_errors.append({"uuid": uuid, "result": {"errors": str(e)}})
```

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py [hunk 4]
```diff
@@ -315,30 +324,48 @@ def handle_upsert_rollback(
...
-        for uuid in delete_objects:
+        return rollback_errors, delete_objects
+
+    def handle_successful_upsert(
+        self, objects_to_remove: list, class_name: str, verbose: bool, tenant: str | None = None
+    ) -> list:
+        """
+        Handles removal of previous objects after successful upsert.
+
+        :param class_name: Name of the class in Weaviate.
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:347]: I would suggest we make the docstring order consistent with the arguments

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py [hunk 5]
```diff
@@ -392,28 +423,49 @@ def ingest_data(
...
+            tenant=tenant,
         )
 
-        if existing == "upsert" and batch_errors:
-            self.logger.warning("Error during upsert. Rolling back all inserts for docs with errors.")
-            rollback_errors = self.handle_upsert_rollback(
-                objects_to_upsert=objects_to_upsert, batch_errors=batch_errors, class_name=class_name, verbose=verbose
-            )
+        if existing == "upsert":
+            if len(self.batch_errors) > 0:
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:438]:
```suggestion
            if self.batch_errors:
```

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py [hunk 6]
```diff
@@ -392,28 +423,49 @@ def ingest_data(
...
+                deletion_errors = self.handle_successful_upsert(
+                    objects_to_remove=objects_to_remove, class_name=class_name, verbose=verbose
+                )
 
-            if len(rollback_errors) > 0:
-                self.logger.error("Errors encountered during rollback.")
-                raise AirflowException("Errors encountered during rollback.")
+                rollback_errors += deletion_errors
+
+                if len(rollback_errors) > 0:
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:450]:
```suggestion
                if rollback_errors:
```

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py [hunk 7]
```diff
@@ -392,28 +423,49 @@ def ingest_data(
...
+                    self.logger.error("Errors encountered during rollback.")
+                    [self.logger.error(f"{rollback_error}" for rollback_error in rollback_errors)]
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:452]:
```suggestion
                    self.logger.error("\n".join(rollback_errors))
```

not sure whether it makes more sense

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py [hunk 8]
```diff
@@ -392,28 +423,49 @@ def ingest_data(
...
+            )
+            if removal_errors:
+                self.logger.error("Errors encountered during removal.")
+                [self.logger.error(f"{removal_error}" for removal_error in removal_errors)]
+                raise AirflowException("Errors encountered during removal.")
 
-        if batch_errors:
+        if self.batch_errors:
             self.logger.error("Errors encountered during ingest.")
+            [self.logger.error(f"{batch_error}" for batch_error in self.batch_errors)]
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:468]: same as above

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py (hunk 1 above)
Comment (sunank200) [reply airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:227]: Done

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py (hunk 2 above)
Comment (sunank200) [reply airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:245]: Done

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py (hunk 3 above)
Comment (sunank200) [reply airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:263]: Done

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py (hunk 5 above)
Comment (sunank200) [reply airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:438]: Done

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py (hunk 4 above)
Comment (sunank200) [reply airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:347]: Done

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py (hunk 6 above)
Comment (sunank200) [reply airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:450]: Done

File: airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py (hunk 8 above)
Comment (sunank200) [reply airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py:468]: Done

**Pull request merged**